app/routes/_layout+/kits.$kitId.give-custody.tsx

```diff
@@ -0,0 +1,269 @@
+      extraInclude: {
+        assets: { select: { status: true } },
+      },
+    });
+
+    if (kit.custody) {
+      return redirect(`/kits/${kitId}`);
+    }
+
+    /**
+     * If any asset is not available in a kit,
+     * then a kit cannot be assigned a custody
+     */
+    const someUnavailableAsset = kit.assets.some(
+      (asset) => asset.status !== "AVAILABLE"
+    );
+    if (someUnavailableAsset) {
+      sendNotification({
+        title: "Cannot assign custody at this time.",
+        message: "One of the asset in kit is not available",
+        icon: { name: "trash", variant: "error" },
+        senderId: userId,
+      });
+
+      return redirect(`/kits/${kitId}`);
+    }
+
+    const searchParams = getCurrentSearchParams(request);
+
+    const teamMembers = await db.teamMember
+      .findMany({
+        where: { deletedAt: null, organizationId },
+        include: { user: true },
+        orderBy: { userId: "asc" },
+        take: searchParams.get("getAll") === "teamMember" ? undefined : 12,
+      })
+      .catch((cause) => {
+        throw new ShelfError({
+          cause,
+          message:
+            "Something went wrong while fetching team members. Please try again or contact support.",
+          label: "Kit",
+        });
+      });
+
+    const totalTeamMembers = await db.teamMember.count({
+      where: { deletedAt: null, organizationId },
+    });
+
+    return json(
+      data({
+        showModal: true,
+        kit,
+        teamMembers,
+        totalTeamMembers,
+      })
+    );
+  } catch (cause) {
+    const reason = makeShelfError(cause, { userId, kitId });
+    return json(error(reason), { status: reason.status });
+  }
+}
+
+export async function action({ context, request, params }: ActionFunctionArgs) {
+  const authSession = context.getSession();
+  const { userId } = authSession;
+
+  const { kitId } = getParams(params, z.object({ kitId: z.string() }), {
+    additionalData: { userId },
+  });
+
+  try {
+    assertIsPost(request);
+
+    await requirePermission({
+      userId,
+      request,
+      entity: PermissionEntity.asset,
+      action: PermissionAction.update,
+    });
+
+    const { custodian } = parseData(
+      await request.formData(),
+      z.object({
+        custodian: stringToJSONSchema.pipe(
+          z.object({ id: z.string(), name: z.string() })
+        ),
+      }),
+      { additionalData: { userId, kitId } }
+    );
+
+    const { id: custodianId, name: custodianName } = custodian;
+
+    const user = await getUserByID(userId);
+
+    const kit = await db.kit.update({
+      where: { id: kitId },
+      data: {
+        status: KitStatus.IN_CUSTODY,
+        custody: { create: { custodian: { connect: { id: custodianId } } } },
+      },
+      include: {
+        assets: true,
+      },
+    });
+
+    await Promise.all([
+      /**
+       * Assign custody to all assets of kit
+       */
+      ...kit.assets.map((asset) =>
+        db.asset.update({
```

Comment:
I would prefer if we use `updateMany`. Doing updates like this in a loop can get very slow and put big strain server.